lib/experimental/sync/class-gutenberg-http-signaling-server.php

```diff
@@ -0,0 +1,380 @@
+		static::save_contents_to_file_descriptor( $fd, $content );
+		flock( $fd, LOCK_UN );
+		fclose( $fd );
+	}
+
+	/**
+	 * Makes the file descriptor content of $fd equal to the serialization of content.
+	 * Overwrites what was previously in $fd.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param resource $fd      A file descriptor.
+	 * @param array    $content Content to be serialized and written in a file descriptor.
+	 */
+	private static function save_contents_to_file_descriptor( $fd, $content ) {
+		rewind( $fd );
+		$data = serialize( $content );
+		fwrite( $fd, $data );
+		ftruncate( $fd, strlen( $data ) );
+	}
+
+	/**
+	 * Handles a wp_ajax signaling server request of client that wants to retrieve its messages.
+	 *
+	 * It returns the client a response following the
+	 * {@link https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format Event stream format}.
+	 *
+	 * ```
+	 * id: <Event ID>
+	 * retry: <Reconnection time, in ms>
+	 * event: <The type of event>
+	 * data: <The message to be sent>
+	 * ```
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param string $subscriber_to_messages_path The path to the file that contains the messages of the subscribers.
+	 * @param string $subscriber_id               The subscriber id.
+	 */
+	private static function handle_read_pending_messages( $subscriber_to_messages_path, $subscriber_id ) {
+		header( 'Content-Type: text/event-stream' );
+		header( 'Cache-Control: no-cache' );
+		list( $fd, $subscriber_to_messages ) = static::get_contents_and_lock_file( $subscriber_to_messages_path );
+		if ( ! $fd ) {
+			$retries = isset( $_COOKIE['signaling_server_retries'] ) ? intval( $_COOKIE['signaling_server_retries'] ) : 0;
+			$secure  = ( 'https' === parse_url( home_url(), PHP_URL_SCHEME ) );
+			setcookie( 'signaling_server_retries', $retries + 1, time() + DAY_IN_SECONDS, SITECOOKIEPATH, '', $secure );
+			echo 'id: ' . time() . PHP_EOL;
+			echo 'event: error' . PHP_EOL;
+			echo 'data: ' . 'Could not open required file.' . PHP_EOL . PHP_EOL;
+			echo 'retry: ' . 3000 * pow( 2, $retries ) . PHP_EOL;
+			exit;
+		}
+
+		if ( isset( $_COOKIE['signaling_server_retries'] ) ) {
+			$secure = ( 'https' === parse_url( home_url(), PHP_URL_SCHEME ) );
+			// unset the cookie using a past expiration date.
+			setcookie( 'signaling_server_retries', 0, time() - DAY_IN_SECONDS, SITECOOKIEPATH, '', $secure );
+		}
+
+		echo 'retry: 3000' . PHP_EOL;
+		if ( isset( $subscriber_to_messages[ $subscriber_id ] ) && count( $subscriber_to_messages[ $subscriber_id ] ) > 0 ) {
+			echo 'id: ' . time() . PHP_EOL;
+			echo 'event: message' . PHP_EOL;
+			echo 'data: ' . wp_json_encode( $subscriber_to_messages[ $subscriber_id ] ) . PHP_EOL . PHP_EOL;
+			$subscriber_to_messages[ $subscriber_id ] = array();
+			static::save_contents_to_file_descriptor( $fd, $subscriber_to_messages );
+		} else {
+			echo PHP_EOL;
```

Comment:
In this code path, the client doesn't have any message to read. And we end up sending `retry: 3000` back to the client. Is that fine? Should we also send `id` and something in `event`, and `data`?

Reply:
I was checking the documentation to update it and realized that it works as before. Forget about this.